packages/util/src/serialize.ts

```diff
@@ -0,0 +1,139 @@
+export type TFunctor<T, C> = (w: C) => T;
+export type TOrTFunctor<T, C> = T | TFunctor<T, C>;
+const createTFunctor = <T, C>(_: TOrTFunctor<T, C>): TFunctor<T, C> => typeof _ === "function" ? _ as TFunctor<T, C> : () => _;
+
+const META = "__META__";
+const meta = "__meta__";
+const to_META = (key: string) => META + key;
+const to_meta = (key: string) => meta + key;
+const from_META = (key: string) => key.substr(META.length);
+const is_META = (key: string) => key.indexOf("__META__") === 0;
+
+/**
+ * Public property call signature
+ */
+export type publish<T, C> = {
+    (): T,
+    (_: TOrTFunctor<T, C>): C
+    reset(): void;
+    modified(): boolean;
+    exists(): boolean;
+};
+
+/**
+ *  Annotation to enable serialization and publish properties on any given class
+ *  Usage:
+ *      @serializable("chart_Column")
+ *      class Column extends LiteWidget {
+ *          ...
+ *      }
+ */
+export function serializable(classID: string) {
+    // tslint:disable-next-line: callable-types
+    return function <T extends { new(...args: any[]): {} }>(constructor: T) {
+        const __proto__ = constructor.prototype;
+        return class extends constructor {
+            constructor(...args: any[]) {
+                super(...args);
+                Object.getOwnPropertyNames(__proto__).filter(is_META).forEach(METAProp => {
+                    const prop = from_META(METAProp);
+                    const metaProp = to_meta(prop);
+                    if ((this as any)[prop]) {
+                        (this as any)[prop] = (this as any)[prop].bind(this, __proto__[METAProp].func);
+                    } else {
+                        (this as any)[prop] = __proto__[METAProp].func.bind(this);
+                    }
+                    (this as any)[prop].reset = __proto__[METAProp].reset.bind(this, metaProp, METAProp);
+                    (this as any)[prop].modified = __proto__[METAProp].modified.bind(this, metaProp, METAProp);
+                    (this as any)[prop].exists = __proto__[METAProp].exists.bind(this, metaProp, METAProp);
+                });
```

Comment:
This loop is my main objection to this code - but I believe it is still better then the current implementations (I am trying to convince myself its ok...).

Reply:
The use of lower and upper case "META" is confusing and makes this difficult to read.